DependencyInjectionContainer/DIContainer.cs

```diff
@@ -1,130 +1,162 @@
-                                            })
-                                            .ToArray();
+            var resolveLocal = (ResolveSource resSource) =>
+                    resolvedServices.AddRange(Services.Where(service => IsServiceOfGivenType(service, typeof(TypeToResolve)))
+                                            .Select(service => (TypeToResolve)InvokeGenericResolve(service.ImplementationType, this, resSource)));
 
-            return resolvedServices;
-        }
+            var resolveNonLocal = (ResolveSource resSource) =>
+            { if (containerParent != null) { resolvedServices.AddRange(containerParent.ResolveMany<TypeToResolve>(resSource)); } };
 
-        private void ThrowExceptionIfServiceWithTypeExists(Type implementationType)
-        {
-            bool containsServiceWithType = services.Any(service => service.ImplementationType == implementationType);
-            if (containsServiceWithType)
+            switch (resolveSource)
             {
-                throw new ArgumentException($"Service with type {implementationType.FullName} has been already registered");
+                case ResolveSource.Any:
+                    resolveLocal(resolveSource);
+                    resolveNonLocal(resolveSource);
+                    break;
+                case ResolveSource.Local:
+                    resolveLocal(resolveSource);
+                    break;
+                case ResolveSource.NonLocal:
+                    if (containerParent == null)
+                        throw new NullReferenceException("This container does not have a parent");
+                    resolveNonLocal(ResolveSource.Any);
+                    break;
+                default:
+                    throw new ArgumentException("Wrong resolve source");
+                    break;
             }
+
+            return resolvedServices;
         }
+
         private bool IsServiceOfGivenType(Service service, Type type) => service.InterfaceType == type ||
             service.ImplementationType == type;
 
-        internal sealed record Service
+        private object InvokeResolveManyForIEnumerable(Type typeWithGeneric, object invokeFrom, ResolveSource resolveSource)
         {
-            public Service(Type implementationType, ServiceLifetime lifetime, Type interfaceType = null)
+            try
+            {
+                var typeToResolve = typeWithGeneric.GetGenericArguments()[0];
+                var resolveManyGeneric = typeof(DIContainer).GetMethod("ResolveMany").MakeGenericMethod(new[] { typeToResolve });
+                return resolveManyGeneric.Invoke(invokeFrom, new object[] { resolveSource });
+            }
+            catch(TargetInvocationException ex)
             {
-                InterfaceType = interfaceType;
-                ImplementationType = implementationType;
-                Lifetime = lifetime;
+                throw ex.InnerException;
```

Comment:
SUPER CRITICAL MISTAKE

https://asusualcoding.wordpress.com/2018/06/02/throw-vs-throwex-and-best-practices-in-c/
https://kumarashwinhubert.com/exception-handling-in-csharp-throw-or-throw-ex

Find more resources if needed until clearly understand why is so critical. Make for yourself a rule how to handle exceptions